hera_pspec/pspecdata.py

```diff
@@ -330,44 +359,52 @@ def q_hat(self, key1, key2, use_identity=False, use_fft=True,
         assert isinstance(key1, tuple)
         assert isinstance(key2, tuple)
         
-        # Whether to use look-up fn. for identity or inverse covariance matrix
-        icov_fn = self.I if use_identity else self.iC
-        
-        # Calculate C^-1 x_1 and C^-1 x_2
+        # Calculate R x_1 and R x_2
         #iC1x, iC2x = 0, 0
         #for _k in k1: iC1x += np.dot(icov_fn(_k), self.x(_k))
         #for _k in k2: iC2x += np.dot(icov_fn(_k), self.x(_k))
-        iC1x = np.dot(icov_fn(key1), self.x(key1))
-        iC2x = np.dot(icov_fn(key2), self.x(key2))
+        Rx1 = np.dot(self._R(key1), self.x(key1))
+        Rx2 = np.dot(self._R(key2), self.x(key2))
             
         # Whether to use FFT or slow direct method
         if use_fft:
-            _iC1x = np.fft.fft(iC1x.conj(), axis=0)
-            _iC2x = np.fft.fft(iC2x.conj(), axis=0)
-            
-            # FIXME: Should include window function (see get_Q)
-            # win = aipy.dsp.gen_window(nk, taper)
-            
+
+            if not taper == 'none':
+                tapering_fct = aipy.dsp.gen_window(self.Nfreqs, taper)
+                Rx1 *= tapering_fct
+                Rx2 *= tapering_fct
+
+            _Rx1 = np.fft.fft(Rx1.conj(), axis=0)
+            _Rx2 = np.fft.fft(Rx2.conj(), axis=0)
+          
             # Conjugated because inconsistent with pspec_cov_v003 otherwise
             # FIXME: Check that this should actually be conjugated
-            return np.conj(  np.fft.fftshift(_iC1x, axes=0).conj() 
-                           * np.fft.fftshift(_iC2x, axes=0) )
+            return np.conj(  np.fft.fftshift(_Rx1, axes=0).conj() 
+                           * np.fft.fftshift(_Rx2, axes=0) )
         else:
             # Slow method, used to explicitly cross-check FFT code
             q = []
             for i in xrange(self.Nfreqs):
-                Q = self.get_Q(i, self.Nfreqs, taper=taper)
-                iCQiC = np.einsum('ab,bc,cd', iC1.T.conj(), Q, iC2) # C^-1 Q C^-1
-                qi = np.sum(self.x(k1).conj() * np.dot(iCQiC, self.x(k2)), axis=0)
+                RQR = np.einsum('ab,bc,cd', self._R(key1).T.conj(), self._Q[i], \
+                                                        self._R(key2)) # R Q R
+                qi = np.sum(self.x(k1).conj()*np.dot(RQR, self.x(k2)), axis=0)
                 q.append(qi)
             return np.array(q)
-    
-    def get_F(self, key1, key2, use_identity=False, true_fisher=False, 
-              taper='none'):
+
+    def get_G(self, key1, key2):
```

Comment:
This gets rid of the FFT method, which is presumably faster. Do we want to take that hit?

Reply:
I don't think get_F ever had an FFT method. It was just the q_hat function that had the FFT option. I think it's possible to write a get_F function (or now get_G function) that uses the FFT method, but that would be a new implementation.